httpclient/request_id_round_tripper.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright © 2024 Acronis International GmbH.
+
+Released under MIT license.
+*/
+
+package httpclient
+
+import (
+	"context"
+	"github.com/acronis/go-appkit/httpserver/middleware"
+	"net/http"
+)
+
+// RequestIDRoundTripper for X-Request-ID header to the request.
+type RequestIDRoundTripper struct {
+	// Delegate is the next RoundTripper in the chain.
+	Delegate http.RoundTripper
+
+	// Opts are the options for the request ID round tripper.
+	Opts RequestIDRoundTripperOpts
+}
+
+// RequestIDRoundTripperOpts for X-Request-ID header to the request options.
+type RequestIDRoundTripperOpts struct {
+	// RequestIDProvider is a function that provides a request ID.
+	RequestIDProvider func(ctx context.Context) string
+}
+
+// NewRequestIDRoundTripper creates an HTTP transport with X-Request-ID header support.
+func NewRequestIDRoundTripper(delegate http.RoundTripper) http.RoundTripper {
+	return &RequestIDRoundTripper{
+		Delegate: delegate,
+	}
+}
+
+// NewRequestIDRoundTripperWithOpts creates an HTTP transport with X-Request-ID header support with options.
+func NewRequestIDRoundTripperWithOpts(delegate http.RoundTripper, opts RequestIDRoundTripperOpts) http.RoundTripper {
+	return &RequestIDRoundTripper{
+		Delegate: delegate,
+		Opts:     opts,
+	}
+}
+
+// getRequestIDProvider returns a function with the request ID provider.
+func (rt *RequestIDRoundTripper) getRequestIDProvider() func(ctx context.Context) string {
+	if rt.Opts.RequestIDProvider != nil {
+		return rt.Opts.RequestIDProvider
+	}
+
+	return middleware.GetRequestIDFromContext
+}
+
+// RoundTrip adds X-Request-ID header to the request.
+func (rt *RequestIDRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	requestID := rt.getRequestIDProvider()(r.Context())
```

Comment:
Small suggestion: let's call rt.getRequestIDProvider() only if `X-Request-ID` header is absent.